### PR TITLE
Reinstate PARAMETER definitions (may duplicate global hash value definitions)

### PR DESCRIPTION
**Pull request by vdonaldson**
Some PARAMETERs are referenced by value, and some by name.
The by-value instance is the primary instance; link all cases to it.

### CODE REVIEW

File: flang/lib/Lower/ConvertVariable.cpp [hunk 1]
```diff
@@ -222,6 +222,46 @@ genInitialDataTarget(Fortran::lower::AbstractConverter &converter,
...
+}
+template <Fortran::common::TypeCategory TC>
+std::string
+getHashName(Fortran::lower::FirOpBuilder &builder,
+            const Fortran::evaluate::Expr<Fortran::evaluate::SomeKind<TC>> &x) {
+  return std::visit([&](const auto &e) { return getHashName(builder, e); },
+                    x.u);
+}
+std::string
+getHashName(Fortran::lower::FirOpBuilder &builder,
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertVariable.cpp:252]: I suspect all 4 of these could be static.

File: flang/lib/Lower/ConvertVariable.cpp [hunk 2]
```diff
@@ -321,6 +369,8 @@ static void instantiateGlobal(Fortran::lower::AbstractConverter &converter,
     if (var.isDeclaration()) {
       // Using a global from a module not defined in this compilation unit.
       mlir::StringAttr externalLinkage;
+      if (sym.attrs().test(Fortran::semantics::Attr::PARAMETER))
+        return; // PARAMETER declarations are never referenced.
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertVariable.cpp:373]: I think we're already lowering the same bits under different hash names and doing so intentionally.
```fortran
   ,,, [ 1, 2, 3, 4 ] ...
```
might be a `4 x i32` or a `2 x 2 x i32`, which get different names for the same read-only bits.

We could maybe try the `linkonce` trick here. Not sure how everyone feels about that though.

[the remark this answers is not in the document]
Comment (vdonaldson) [reply flang/lib/Lower/ConvertVariable.cpp]: Thanks for pointing this out Jean.  I've reverted the code to typically generate both named and hashed constant definition.

**Review by jeanPerier — APPROVED** [review]
LGTM, if MLIR support alias at some point, it should not be too hard to revisit this.

**Review by schweitzpgi — APPROVED** [review]
Thanks for fixing this, Val,

**Pull request merged**